src/lsm/groove.zig

```diff
@@ -886,37 +888,55 @@ pub fn GrooveType(
             }
         }
 
-        /// Insert the value (or update it, if it exists).
+        /// Update the value. Requires the old object to be provided.
         /// Update the object and index trees by diff'ing the old and new values.
-        pub fn upsert(groove: *Groove, new: *const Object) void {
-            // Explicit stack copy needed - otherwise old will == new!
-            const old: Object = (groove.objects_cache.get(@field(new, primary_field)) orelse {
-                return groove.insert(new);
-            }).*;
+        pub fn update(
+            groove: *Groove,
+            values: struct { old: *const Object, new: *const Object },
+        ) void {
+            const old = values.old;
+            const new = values.new;
 
-            assert(@field(old, primary_field) == @field(new, primary_field));
-            assert(old.timestamp == new.timestamp);
+            if (constants.verify) {
+                const old_from_cache = groove.objects_cache.get(@field(old, primary_field)).?;
+                assert(@field(old_from_cache, primary_field) == @field(old, primary_field));
+                assert(old_from_cache.timestamp == old.timestamp);
+            }
 
-            groove.objects_cache.upsert(new);
+            // Sanity check to ensure the caller didn't accidentally pass in an alias.
+            assert(new != old);
 
-            // The ID can't change, so no need to update the ID tree.
+            if (has_id) {
+                assert(old.id == new.id);
+            }
+            assert(old.timestamp == new.timestamp);
 
-            // Update the object tree entry if any of the fields (even ignored) are different.
-            if (!std.mem.eql(u8, std.mem.asBytes(&old), std.mem.asBytes(new))) {
-                // Unlike the index trees, the new and old values in the object tree share the
-                // same key. Therefore put() is sufficient to overwrite the old value.
-                groove.objects.put(new);
+            // The ID can't change, so no need to update the ID tree. Update the object tree entry
+            // if any of the fields (even ignored) are different. We assume the caller will pass in
+            // an object that has changes.
+            // Unlike the index trees, the new and old values in the object tree share the same
+            // key. Therefore put() is sufficient to overwrite the old value.
+            if (constants.verify) {
+                const tombstone = ObjectTreeHelpers(Object).tombstone;
+                const key_from_value = ObjectTreeHelpers(Object).key_from_value;
+
+                assert(!stdx.equal_bytes(Object, old, new));
+                assert(key_from_value(old) == key_from_value(new));
+                assert(!tombstone(old) and !tombstone(new));
             }
 
+            groove.objects.put(new);
+            groove.objects_cache.upsert(new);
```

Comment:
This makes my [my Rust eye](https://static.wikia.nocookie.net/p__/images/7/73/Mad-Eye_Moody%27s_%22mad_eye%22.jpeg/revision/latest?cb=20160215100736&path-prefix=protagonist) twitch a bit. Isn't this shared mutable aliasing? I _think_ the `old` object might be the object from the cache (`maybe(old == old_from_cache)` in the if above). If that's the case, I think we _might_ be clobbering old here, in which case the `inline for` below would get wrong results when accessing `old` fields?

Reply:
We have an explicit guard against this above; since I've been bitten by this before here :smile::

> assert(new != old);

I think that's enough to prevent it from happening?

Reply:
No, I think that's different. The `new` is an object on state_machine stack:

https://github.com/tigerbeetle/tigerbeetle/blob/c367d1dfdba783b8eb171cdfec5d9de0188517c5/src/state_machine.zig#L870

the `old` is an object from the cache:

https://github.com/tigerbeetle/tigerbeetle/blob/c367d1dfdba783b8eb171cdfec5d9de0188517c5/src/state_machine.zig#L813

So `new` and `old`  are different. 

However, when we do `groove.objects_cache.upsert(new);` this might invalidate `old` still, if it was in the cache. That is, after this line I would expect `new.*` and `old.*` to be equal, because the `old` memory was overwritten when we upserted `new`.

Not 100% sure that this could happen, but I think it does?

Reply:
Ahhh yes, that is the case. Good catch!

In the past, I took an explicit stack copy of `old` within the function which prevented that. We could move the object cache upsert to the end - there's no reason it needs to be done before the index trees - with an explicit comment explaining the above. That way we can get correct behaviour without a stack copy, but it still feels a bit fragile...

Reply:
I think that's ok, as long as we document that with a `maybe`. 

I wonder if we even can _require_ that this is the object from the cache? It _must_ be, if we are updating, right?

Reply:
I am also _reeeeeally_ wondering why CI is green here :thinking: in `release`, I can see compiler hoisting the loads to before the cache update, which would give the correct behavior. 

But in debug, I expect the overwrite to be reflected robustly. Could you also take a look what's up with tests passing here? Either my theory about the overwrite is wrong (good case), or we somehow don't test this case at all (which would be pretty scary).

Reply:
> Could you also take a look what's up with tests passing here? Either my theory about the overwrite is wrong (good case), or we somehow don't test this case at all (which would be pretty scary).

As far as I can tell, we don't have any tests (unit or fuzz) that actually check the result of the secondary index trees, and in this case, if `old.* == new.*` it'll just skip updating the secondary index so there's nothing that's hit on the way down :grimacing:...

Also just noticed the forest fuzz (which I forgot to run / update in this PR) has also been broken for a few weeks. I need to look at proper infra for fuzzing ASAP, but maybe as a workaround for now we can just run each fuzz for a fixed time on CI?

We previously did run some more fuzzes in CI, but with a fixed seed for a single iteration.

Reply:
> I wonder if we even can require that this is the object from the cache? It must be, if we are updating, right?

It could in theory be a stack copy, but we don't have an explicit use case for that yet so I've made the check `assert(old_from_cache == old);`.

Reply:
> Also just noticed the forest fuzz (which I forgot to run / update in this PR) has also been broken for a few weeks.

I think that is fixed as part of https://github.com/tigerbeetle/tigerbeetle/pull/1234, which merged this morning.